### PR TITLE
feat: Magefile

### PR DESCRIPTION
**Pull request by sh0rez**
Adds a `magefile.go` as an exploratory effort whether it is worth it investing on that technology. Not intended to be merged yet. See #684 for reference

Progress overview:
* [ ] `all`
* [x] `assets`
* [ ] `build-plugin`
* [x] `check_assets`
* [ ] `check-generated-files`
* [ ] `clean`
* [ ] `cmd/docker-driver/docker-driver`
* [ ] `cmd/docker-driver/docker-driver-debug`
* [ ] `cmd/docker-driver/.uptodate`
* [ ] `cmd/docker-driver/.uptodate-debug`
* [x] `cmd/logcli/logcli` (renamed to `build:logcli`)
* [x] `cmd/logcli/logcli-debug` (renamed to `DEBUG=true build:logcli`)
* [ ] `cmd/logcli/.uptodate`
* [ ] `cmd/logcli/.uptodate-debug`
* [x] `cmd/loki/loki` (renamed to `build:loki`)
* [x] `cmd/loki/loki-debug` (renamed to `DEBUG=true build:loki`)
* [ ] `cmd/loki/.uptodate`
* [ ] `cmd/loki/.uptodate-debug`
* [x] `cmd/promtail/promtail` (renamed to `build:logcli`)
* [x] `cmd/promtail/promtail-debug` (renamed to `DEBUG=true build:loki`)
* [ ] `cmd/promtail/.uptodate`
* [ ] `cmd/promtail/.uptodate-debug`
* [ ] `debug`
* [ ] `enable-plugin`
* [x] `helm` (renamed to `helm:build`)
* [x] `helm-clean`
* [x] `helm-install`
* [x] `helm-publish`
* [x] `helm-upgrade`
* [ ] `images`
* [x] `lint`
* [ ] `load-images`
* [ ] `loki-build-image/.uptodate`
* [ ] `pkg/logproto/logproto.pb.go`
* [ ] `pkg/logql/expr.go`
* [ ] `pkg/promtail/server/server.go`
* [ ] `protos`
* [ ] `push-images`
* [ ] `push-latest`
* [ ] `push-plugin`
* [ ] `release-perform`
* [ ] `release-prepare`
* [ ] `save-images`
* [ ] `shell`
* [x] `test`
* [ ] `vendor/github.com/cortexproject/cortex/pkg/chunk/storage/caching_index_client.pb.go`
* [ ] `vendor/github.com/cortexproject/cortex/pkg/ingester/client/cortex.pb.go`
* [ ] `vendor/github.com/cortexproject/cortex/pkg/ring/ring.pb.go`
* [ ] `yacc`

### CODE REVIEW

**Comment by slim-bean** [comment]
I'm liking this so far, my only gripe is that I wish you could default the `-v` flag as I like seeing output by default.  

I think we should continue to build out functionality.

**Comment by sh0rez** [comment]
`alias mage=mage - v`?

**Comment by slim-bean** [comment]
yeah true :) really I want to impose `-v` on everyone, which we can do if we create make targets that call mage.  

I think to keep the feel of loki similar to most go projects we should keep the makefile as a wrapper to mage, we can even go so far as adding the new `mage` target as a pre-req in make for `all` so the use of mage could be transparent to people.

I think we'd just have to figure out what targets to keep in the makefile

File: mage/assets.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+type Assets mg.Namespace
+
+func (Assets) CheckAssets() error {
+	err := sh.Run("git", "diff", "--exit-code", "pkg/promtail/server/ui")
```
Comment (sh0rez) [inline mage/assets.go]: `git()` is available as a function (`cmds.go`)

**Comment by jpreese** [comment]
@sh0rez  What was the end result of this experiment? We're contemplating investing in Mage to help with compatibility issues with Makefile and supporting devs on Windows/Max/Linux

**Comment by sh0rez** [comment]
We went with Makefiles instead, mostly because shelling out in Go is not optimal, as it requires error checking, etc.

But I am aware of many projects being very happy with Mage and it is by far the best Make alternative I have seen so far.

**Comment by jpreese** [comment]
Understood, thanks for the quick response!

**Pull request closed without merging**